lighthouse-core/gather/driver.js

```diff
@@ -425,6 +425,66 @@ class Driver {
   }
 
   /**
+   * Installs a PerformanceObserver in the page to monitor for longtasks and resolves when there have
+   * been no long tasks for at least waitForCPUQuiet ms. The promise will not resolve until the
+   * `markAsResolvable` function on the return object has been called. This is to prevent promise resolution
+   * before some important point in time such as network quiet or document load.
+   * @param {number} waitForCPUQuiet
+   * @return {{promise: !Promise, cancel: function(), markAsResolvable: function()}}
+   */
+  _waitForCPUIdle(waitForCPUQuiet) {
+    if (!waitForCPUQuiet) {
+      return {
+        markAsResolvable: () => undefined,
+        promise: Promise.resolve(),
+        cancel: () => undefined,
+      };
+    }
+
+    let lastTimeout;
+    let isResolvable = false;
+    function checkForQuiet(driver, resolve) {
+      const tryLater = () => {
+        lastTimeout = setTimeout(() => checkForQuiet(driver, resolve), 1000);
+      };
+
+      if (!isResolvable) {
+        return tryLater();
+      }
+
+      return driver.evaluateAsync(`(${checkTimeSinceLastLongTask.toString()})()`)
+        .then(timeSinceLongTask => {
+          if (typeof timeSinceLongTask === 'number' && timeSinceLongTask >= waitForCPUQuiet) {
+            log.verbose('Driver', `CPU has been idle for ${timeSinceLongTask} ms`);
+            resolve();
+          } else {
+            log.verbose('Driver', `CPU has been idle for ${timeSinceLongTask} ms`);
+            tryLater();
+          }
+        });
+    }
```

Comment:
I'm not quite as close to this code as @brendankenny or @paulirish might be, but the general approach to using long-tasks here sgtm.